machineconfiguration/v1alpha1/types_machineconfigstate.go

```diff
@@ -0,0 +1,125 @@
+package v1alpha1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +genclient:nonNamespaced
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigState describes the health of the Machines on the system
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineConfigState struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// +kubebuilder:validation:Required
+	Spec MachineConfigStateSpec `json:"spec"`
+	// +optional
+	Status MachineConfigStateStatus `json:"status"`
+}
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// MachineConfigStateList describes all of the MachinesStates on the system
+//
+// Compatibility level 4: No compatibility is provided, the API can change at any point for any reason. These capabilities should not be used by applications needing long term support.
+// +openshift:compatibility-gen:level=4
+type MachineConfigStateList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata"`
+
+	Items []MachineConfigState `json:"items"`
+}
+
+// MachineConfigStateSpec describes the type of State we are managing
+type MachineConfigStateSpec struct {
+	// configuration describes the object reference information of a MachineConfigState
+	Config MachineConfigStateConfig `json:"configuration"`
+}
+
+// MachineConfigStateStatus holds the reported information on a particular MachineConfigState
+type MachineConfigStateStatus struct {
+	corev1.ObjectReference `json:",inline"`
+	// conditions represents the latest available observations of current state.
+	// +listType=atomic
+	// +optional
+	Conditions []ProgressionCondition `json:"conditions"`
+	// mostRecentError is populated if the State reports an error.
+	MostRecentError string `json:"mostRecentError"`
+	// health reports the overall status of this MachineConfigState Given its Progress
+	Health MachineConfigStateHealthEnumeration `json:"health"`
+	// observedGeneration represents the generation observed by the controller.
+	// +optional
+	ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+	// configVersion holds the current and desired config versions
+	ConfigVersion MachineConfigVersion `json:"configVersion"`
+}
+
+// ConfigVersion holds the current and desired config versions
+type MachineConfigVersion struct {
+	// Current is the current MachineConfig a node is using
+	Current string `json:"current"`
+	// Desired is the MachineConfig the node wants to upgrade to.
+	Desired string `json:"desired"`
+}
```

Comment:
Are the fields in this struct optional or required? What validations can be applied? These are names of resources in Kubernetes so they must adhere to the RFC1123 subdomain spec right?
We should add maximum and minimum length requirements and a pattern to validate the entry.
Then the godoc should explain these validations

Reply:
I can make them required as the annotations for these configs always must exist.